src/nsidc/metgen/metgen.py

```diff
@@ -109,100 +142,198 @@ def scrub_json_files(path):
+
+def recorder(fn: Callable[[Granule], Granule], ledger: Ledger) -> Ledger:
+    """
+    Higher-order function that, given a granule operation function and a
+    Ledger, will execute the function on the Ledger's granule, record the
+    results, and return the resulting new Ledger.
+    """
+    # Execute the operation and record the result
+    successful = True
+    message = ''
+    start = dt.datetime.now()
+    try:
+        new_granule = fn(ledger.granule)
+    except Exception as e:
+        successful = False
+        message = str(e)
+    end = dt.datetime.now()
+
+    # Store the result in the Ledger
+    new_actions = ledger.actions.copy()
+    new_actions.append(
+        Action(
+            fn.func.__name__,
+            successful=successful,
+            message=message,
+            startDatetime=start,
+            endDatetime=end
+        )
+    )
+
+    return dataclasses.replace(
+        ledger,
+        granule=new_granule,
+        actions=new_actions
+    )
+
+def start_ledger(granule: Granule) -> Ledger:
+    """
+    Start a new Ledger of the operations on the given Granule.
+    """
+    return Ledger(
+        granule,
+        startDatetime=dt.datetime.now()
+    )
 
-def granule_paths(data_dir):
-    # Returns a list of tuples containing the "producer granule id" and a dict
-    # containing the key 'data' with a list of one or more files identified as 
-    # part of the same granule. We still need code to identify the common
-    # basename for the cases where more than one file exists per granule (or
-    # the case where an ancillary file is associated with the granule), and to
-    # add the correct "type" of the file. See the CNM spec for a list of types.
-    # At the moment the assumption is one (netCDF!) file per granule, with a
-    # type of "data," plus a metadata (UMM-G) file.
-    producer_granule_ids = [os.path.basename(f) for f in data_dir.glob('*.nc')]
-    granule_data_files = [{ 'data': [f] } for f in data_dir.glob('*.nc')]
+def end_ledger(ledger: Ledger) -> Ledger:
+    """
+    Finalize the Ledger of operations on its Granule.
+    """
+    return dataclasses.replace(
+        ledger,
+        endDatetime=dt.datetime.now(),
+        successful=all([a.successful for a in ledger.actions])
+    )
 
-    return list(zip(producer_granule_ids, granule_data_files))
+# -------------------------------------------------------------------
+# Granule Operations
+# -------------------------------------------------------------------
 
-def find_or_create_ummg(mapping, data_file_paths, ummg_path, all_existing_ummg):
+def granule_collection(configuration: config.Config, granule: Granule) -> Granule:
     """
-    Look for an existing UMM-G file. If nothing found, create a new one.
+    Find the Granule's Collection and add it to the Granule.
+    """
+    return dataclasses.replace(
+        granule, 
+        collection=Collection(configuration.auth_id, configuration.version)
```

Comment:
There are two parts to collection information: getting it from CMR (network traffic), and attaching it to the granule operations. This comment is only a reminder that once we start grabbing information from CMR we'll need to separate the "talk to CMR" part (which only needs to happen once) out of this series of granule-specific operations.

Reply:
Oh, good catch!